File: Sources/ActivitiesCore/FileScanner.swift
import Foundation
#if canImport(os)
import os
#endif

/// Ergebnis eines Suchlaufs.
public struct ScanOutcome: Sendable {
    public let files: [RelevantFile]
    /// Ordner, die wegen einer **Namensregel** uebersprungen wurden
    /// (`node_modules`, `.build` …).
    public let skippedByRule: Int
    /// Ordner, die der Anwender **selbst** ausgeblendet hat und die dieser
    /// Suchlauf deshalb nicht betreten hat.
    ///
    /// **⚠️ Getrennt gezaehlt, seit die Kopfzone beide Zahlen nebeneinander
    /// nennt.** Bis v1.19.65 lief beides auf einen Zaehler, und die Kopfzone
    /// schrieb „35 Ordner samt Inhalt uebersprungen · 2 von dir ausgeblendet" –
    /// die 2 steckten in den 35, und die zweite Zahl kam ausserdem aus einer
    /// ganz anderen Quelle (der Anzahl der **Regeln**, nicht der Ordner). Zwei
    /// Zahlen nebeneinander, von denen eine die andere enthaelt, kann niemand
    /// lesen. Jetzt sind sie disjunkt und von derselben Art.
    public let skippedByHiddenPath: Int

    /// Alle uebersprungenen Ordner zusammen.
    public var skippedFolders: Int { skippedByRule + skippedByHiddenPath }

    public init(files: [RelevantFile], skippedByRule: Int = 0, skippedByHiddenPath: Int = 0) {
        self.files = files
        self.skippedByRule = skippedByRule
        self.skippedByHiddenPath = skippedByHiddenPath
    }
}

/// Durchsucht einen Verzeichnisbaum nach kuerzlich bearbeiteten Dateien.
///
/// Ausgewertet werden ausschliesslich Dateien. Das massgebliche Datum je Datei
/// ist das neuere aus Erstell- und Aenderungsdatum. Versteckte Objekte sowie
/// bekannte Junk-Dateien/-Ordner werden ausgeschlossen; Symlinks werden nicht
/// verfolgt. Nicht lesbare Eintraege werden uebersprungen und protokolliert.
public struct FileScanner: Sendable {
    private let exclusions: ExclusionRules
    /// Protokollierung ist plattformabhaengig gekapselt: ``os.Logger`` gibt es
    /// nur auf Apple-Systemen. Der Kern soll ohne Apple-Frameworks uebersetzbar
    /// bleiben (siehe Konzept „Portabilitaet").
    #if canImport(os)
    private static let logger = Logger(subsystem: "com.mtri.activities", category: "scanner")
    #endif

    /// Meldet einen uebersprungenen Eintrag – auf Apple-Systemen ueber
    /// ``os.Logger``, sonst ueber die Standardfehlerausgabe.
    private static func logSkipped(_ url: URL, _ error: Error) {
        #if canImport(os)
        logger.warning("Eintrag uebersprungen (\(url.path, privacy: .public)): \(error.localizedDescription, privacy: .public)")
        #else
        FileHandle.standardError.write(
            Data("Eintrag uebersprungen (\(url.path)): \(error.localizedDescription)\n".utf8)
        )
        #endif
    }

    public init(exclusions: ExclusionRules = .default) {
        self.exclusions = exclusions
    }

    /// Ermittelt das neuere aus Erstell- und Aenderungsdatum.
    func effectiveTimestamp(creation: Date?, modification: Date?) -> Date {
        let created = creation ?? .distantPast
        let modified = modification ?? .distantPast
        return max(created, modified)
    }

    /// Liefert alle Dateien im Wurzelbaum, deren Datum im Intervall ``[start, end)`` liegt.
    ///
    /// - Parameters:
    ///   - settings: Wurzelordner, Zeitfenster (start inklusiv, end exklusiv) und Namensmuster.
    ///   - shouldCancel: Wird regelmaessig geprueft; liefert ``true``, bricht der Scan ab.
    ///   - onProgress: Wird periodisch mit der Zahl der bisher geprueften Eintraege aufgerufen.
    public func scan(
        settings: ScanSettings,
        shouldCancel: () -> Bool = { false },
        onProgress: (Int) -> Void = { _ in }
    ) -> ScanOutcome {
        let filter = NameFilter(settings.namePattern)
        let start = settings.start
        let end = settings.end

        // `.fileSizeKey` kostet nichts: Gemessen ueber 5.266 Dateien in drei
        // Laeufen lag der Aufschlag bei −2,3 % – also im Rauschen. Die Werte
        // holt derselbe Aufruf, der ohnehin stattfindet.
        let keys: Set<URLResourceKey> = [
            .creationDateKey, .contentModificationDateKey, .isDirectoryKey,
            .isRegularFileKey, .isSymbolicLinkKey, .nameKey, .isPackageKey,
            .fileSizeKey,
        ]
        let fileManager = FileManager.default
        guard let enumerator = fileManager.enumerator(
            at: settings.rootURL,
            includingPropertiesForKeys: Array(keys),
            options: [.skipsHiddenFiles],
            errorHandler: { url, error in
                Self.logSkipped(url, error)
                return true
            }
        ) else {
            return ScanOutcome(files: [])
        }

        // **Ein einziges Array, ohne Obergrenze – und das ist gemessen, nicht
        // geraten.**
        //
        // Der Verdacht aus PR-25 lautete: Bei einem sehr grossen Wurzelordner
        // waechst diese Liste ungebremst, bis das Fenster steht. `swift run -c
        // release Bench --disk 500000` sagt dazu (M-Chip, APFS, warmer Cache):
        //
        //     500.000 Dateien   Suchlauf 10,0 s   Spitze +550 MB
        //     Abbruch nach 1000 Eintraegen: 23 ms
        //
        // Damit ist eine Obergrenze **nicht** gerechtfertigt. Der Suchlauf
        // laeuft neben der Oberflaeche, und ``shouldCancel`` greift in 23 ms –
        // der Knopf „Abbrechen" ist also keine Behauptung, sondern haelt auch
        // beim Zwanzigfachen des bisher gemessenen Bestandes. Ein Deckel
        // wuerde hier Arbeit verweigern, die das Programm nachweislich leistet.
        //
        // ⚠️ Wer spaeter doch einen Deckel einzieht, muss eine Zahl mitbringen,
        // die diese widerlegt – nicht ein Gefuehl von Vorsicht.
        var results: [RelevantFile] = []
        var examined = 0
        // Zaehlt uebersprungene Ordner, damit die App offenlegen kann, wie viel
        // sie ausblendet – Ausschluesse duerfen kein stiller Zustand sein.
        // Getrennt nach Grund, weil die Kopfzone beide Zahlen nebeneinander
        // nennt und sie sich dann nicht ueberschneiden duerfen.
        var skippedByRule = 0
        var skippedByHiddenPath = 0
        for case let fileURL as URL in enumerator {
            if shouldCancel() { break }
            examined += 1
            if examined & 1023 == 0 { onProgress(examined) }

            guard let values = try? fileURL.resourceValues(forKeys: keys) else { continue }
            let name = values.name ?? fileURL.lastPathComponent

            // Symlinks nicht verfolgen.
            if values.isSymbolicLink == true {
                if values.isDirectory == true { enumerator.skipDescendants() }
                continue
            }

            if values.isDirectory == true {
                // Ausgeschlossene Ordner und ausgeblendete Pfade nicht betreten.
                if exclusions.isExcludedFolder(name) {
                    skippedByRule += 1
                    enumerator.skipDescendants()
                    continue
                }
                if exclusions.isExcludedPath(fileURL.path) {
                    skippedByHiddenPath += 1
                    enumerator.skipDescendants()
                    continue
                }

                // **App-Buendel sind Dokumente, keine Ordner.** Ohne diese
                // Behandlung meldete die App deren Innereien
                // (`.../Contents/_CodeSignature`) als „bearbeitete Ordner" –
                // dort hat nie ein Mensch gearbeitet.
                let isPackage = values.isPackage
                    ?? ExclusionRules.isPackage(extension: fileURL.pathExtension)
                if isPackage {
                    enumerator.skipDescendants()
                    if !exclusions.isExcludedFile(name), filter.matches(name) {
                        let timestamp = effectiveTimestamp(
                            creation: values.creationDate,
                            modification: values.contentModificationDate
                        )
                        if timestamp >= start && timestamp < end {
                            results.append(
                                RelevantFile(
                                    url: fileURL,
                                    folder: fileURL.deletingLastPathComponent(),
                                    timestamp: timestamp,
                                    size: values.fileSize
                                )
                            )
                        }
                    }
                }
                continue
            }

            guard values.isRegularFile == true else { continue }
            if exclusions.isExcludedFile(name) { continue }
            if !filter.matches(name) { continue }

            let timestamp = effectiveTimestamp(
                creation: values.creationDate,
                modification: values.contentModificationDate
            )
            if timestamp >= start && timestamp < end {
                results.append(
                    RelevantFile(
                        url: fileURL,
                        folder: fileURL.deletingLastPathComponent(),
                        timestamp: timestamp,
                        size: values.fileSize
                    )
                )
            }
        }
        onProgress(examined)
        return ScanOutcome(
            files: results,
            skippedByRule: skippedByRule,
            skippedByHiddenPath: skippedByHiddenPath
        )
    }

    /// Listet die Dateien direkt im Ordner - ohne Zeitraumgrenze, aber mit
    /// Namensfilter (fuer die aufklappbare Detailansicht). Ergebnis nach Datum
    /// absteigend (bei Gleichstand alphabetisch).
    public func listDirectoryFiles(_ folder: URL, filter: NameFilter) -> [RelevantFile] {
        let keys: Set<URLResourceKey> = [
            .creationDateKey, .contentModificationDateKey,
            .isRegularFileKey, .isSymbolicLinkKey, .nameKey, .fileSizeKey,
        ]
        let fileManager = FileManager.default
        guard let contents = try? fileManager.contentsOfDirectory(
            at: folder,
            includingPropertiesForKeys: Array(keys),
            options: [.skipsHiddenFiles]
        ) else {
            return []
        }

        var files: [RelevantFile] = []
        for url in contents {
            guard let values = try? url.resourceValues(forKeys: keys) else { continue }
            let name = values.name ?? url.lastPathComponent
            if values.isSymbolicLink == true { continue }
            guard values.isRegularFile == true else { continue }
            if exclusions.isExcludedFile(name) { continue }
            if !filter.matches(name) { continue }
            files.append(
                RelevantFile(
                    url: url,
                    folder: folder,
                    timestamp: effectiveTimestamp(
                        creation: values.creationDate,
                        modification: values.contentModificationDate
                    ),
                    size: values.fileSize
                )
            )
        }

        files.sort { first, second in
            if first.timestamp != second.timestamp {
                return first.timestamp > second.timestamp
            }
            return first.url.lastPathComponent.localizedCaseInsensitiveCompare(second.url.lastPathComponent) == .orderedAscending
        }
        return files
    }
}

File: legacy-python-cli/recent_files/file_scanner.py
"""Durchsucht den Verzeichnisbaum nach kuerzlich bearbeiteten Dateien.

Ausgewertet werden ausschliesslich Dateien. Das massgebliche Datum je Datei ist
das neuere aus Erstell- und Aenderungsdatum. Versteckte Objekte sowie bekannte
Junk-Dateien/-Ordner werden ausgeschlossen; Symlinks werden nicht verfolgt.
"""

from __future__ import annotations

import logging
import os
import stat as stat_module
from dataclasses import dataclass
from datetime import datetime, timedelta
from fnmatch import fnmatch
from pathlib import Path
from typing import Iterator

from recent_files.configuration import Configuration

logger = logging.getLogger(__name__)

# Windows-Attribut fuer versteckte Dateien; auf anderen Systemen nicht vorhanden.
_FILE_ATTRIBUTE_HIDDEN = getattr(stat_module, "FILE_ATTRIBUTE_HIDDEN", 0x2)


@dataclass(frozen=True)
class RelevantFile:
    """Eine im Zeitraum liegende Datei samt ihrem beinhaltenden Ordner."""

    path: Path
    folder: Path
    timestamp: datetime


def effective_timestamp(stat_result: os.stat_result) -> datetime:
    """Ermittelt das neuere aus Erstell- und Aenderungsdatum.

    Das Erstelldatum ist betriebssystemabhaengig: macOS nutzt ``st_birthtime``,
    Windows ``st_ctime``. Fehlt ``st_birthtime``, dient ``st_ctime`` als Ersatz.
    """
    created = getattr(stat_result, "st_birthtime", stat_result.st_ctime)
    modified = stat_result.st_mtime
    return datetime.fromtimestamp(max(created, modified))


def is_hidden(name: str, stat_result: os.stat_result | None = None) -> bool:
    """Prueft, ob ein Eintrag versteckt ist (Dotfile oder Windows-Attribut)."""
    if name.startswith("."):
        return True
    attributes = getattr(stat_result, "st_file_attributes", 0) if stat_result else 0
    return bool(attributes & _FILE_ATTRIBUTE_HIDDEN)


def is_excluded_file(name: str, excluded_files: frozenset[str]) -> bool:
    """Prueft einen Dateinamen gegen die Ausschlussmuster (inkl. Glob wie ``~$*``)."""
    return any(fnmatch(name, pattern) for pattern in excluded_files)


def matches_extension_filter(name: str, extension_filters: frozenset[str]) -> bool:
    """Prueft, ob die Endung einer Datei einem der Filter-Muster entspricht.

    Ist kein Filter gesetzt, gilt jede Datei als passend. Andernfalls wird die
    Endung (klein, ohne Punkt) per Glob gegen die Muster geprueft (``xls*`` passt
    auf ``xls`` und ``xlsx``).
    """
    if not extension_filters:
        return True
    extension = Path(name).suffix.lower().lstrip(".")
    return any(fnmatch(extension, pattern) for pattern in extension_filters)


def find_relevant_files(config: Configuration) -> Iterator[RelevantFile]:
    """Liefert alle Dateien im Wurzelbaum, deren Datum im Zeitraum liegt.

    Ausgeschlossene Ordner werden vor dem Abstieg entfernt. Nicht lesbare
    Eintraege werden uebersprungen und protokolliert; Symlinks werden nicht
    verfolgt.
    """
    cutoff = datetime.now() - timedelta(days=config.days)

    for current_dir, dir_names, file_names in os.walk(config.root_path, topdown=True):
        current_path = Path(current_dir)

        # Ausgeschlossene und versteckte Unterordner nicht betreten.
        dir_names[:] = [
            name
            for name in dir_names
            if name not in config.excluded_folders and not is_hidden(name)
        ]

        for file_name in file_names:
            if is_hidden(file_name) or is_excluded_file(file_name, config.excluded_files):
                continue
            if not matches_extension_filter(file_name, config.extension_filters):
                continue

            file_path = current_path / file_name
            if file_path.is_symlink():
                continue

            try:
                stat_result = file_path.stat()
            except OSError as error:
                logger.warning("Datei uebersprungen (%s): %s", file_path, error)
                continue

            timestamp = effective_timestamp(stat_result)
            if timestamp >= cutoff:
                yield RelevantFile(
                    path=file_path,
                    folder=current_path,
                    timestamp=timestamp,
                )


def list_directory_files(folder: Path, excluded_files: frozenset[str]) -> list[RelevantFile]:
    """Listet alle Dateien direkt im Ordner - ohne Zeitraum- oder Endungsfilter.

    Gedacht fuer die aufklappbare Detailansicht: es werden auch aeltere Dateien
    angezeigt. Versteckte Objekte und bekannte Junk-Dateien werden ausgelassen,
    Symlinks nicht verfolgt. Ergebnis nach Datum absteigend (juengste zuerst,
    bei Gleichstand alphabetisch).
    """
    files: list[RelevantFile] = []
    try:
        with os.scandir(folder) as iterator:
            for dir_entry in iterator:
                name = dir_entry.name
                if is_hidden(name) or is_excluded_file(name, excluded_files):
                    continue
                try:
                    if dir_entry.is_symlink() or not dir_entry.is_file():
                        continue
                    stat_result = dir_entry.stat()
                except OSError as error:
                    logger.warning("Datei uebersprungen (%s): %s", name, error)
                    continue
                files.append(
                    RelevantFile(
                        path=Path(dir_entry.path),
                        folder=folder,
                        timestamp=effective_timestamp(stat_result),
                    )
                )
    except OSError as error:
        logger.warning("Ordner nicht lesbar (%s): %s", folder, error)
        return []

    files.sort(key=lambda relevant_file: relevant_file.path.name.lower())
    files.sort(key=lambda relevant_file: relevant_file.timestamp, reverse=True)
    return files
